graphs/dfs.py

```diff
@@ -1,33 +1,34 @@
-"""pseudo-code"""
-
-"""
-DFS(graph G, start vertex s):
-// all nodes initially unexplored
-mark s as explored
-for every edge (s, v):
-    if v unexplored:
-        DFS(G, v)
+"""The DFS function simply calls itself recursively for every unvisited child of 
+its argument. We can emulate that behaviour precisely using a stack of iterators. 
+Instead of recursively calling with a node, we'll push an iterator to the node's 
+children onto the iterator stack. When the iterator at the top of the stack
+terminates, we'll pop it off the stack.
+
+Pseudocode:
+    all nodes initially unexplored
+    mark s as explored
+    for every edge (s, v):
+        if v unexplored:
+            DFS(G, v)
 """
 
+from typing import Set, Dict
 
-def dfs(graph, start):
-    """The DFS function simply calls itself recursively for every unvisited child of its argument. We can emulate that
-     behaviour precisely using a stack of iterators. Instead of recursively calling with a node, we'll push an iterator
-      to the node's children onto the iterator stack. When the iterator at the top of the stack terminates, we'll pop
-       it off the stack."""
-    explored, stack = set(), [start]
-    while stack:
-        v = (
-            stack.pop()
-        )  # one difference from BFS is to pop last element here instead of first one
 
-        if v in explored:
-            continue
-
-        explored.add(v)
+def dfs(graph: Dict, start: str) -> Set[int]:
+    """Depth First Search on Graph
 
+       :param graph: directed graph in dictionary format
+       :param vertex: starting vectex as a string
+       :returns: the trace of the search
```

Comment:
Please add a doctest.

Reply:
I'm not sure how to write doctest for this one. The function returns different outputs on different runs.

Reply:
Could we do:
```
>>> all(x in output for x in graph)
True
# and/or
>>> all(x in output for x in graph.values())
True
```

Reply:
If not then perhaps add a comment explaining that doctests would be difficult.

Reply:
That works!